DOC: Added a PR "Reviewer guidelines" document.

melissawm opened the pull request:
This is a first draft. You may notice some parts are based off the [SciPy reviewer guide](https://docs.scipy.org/doc/scipy/reference/dev/contributor/reviewing_prs.html), and here's some other projects guidelines for comparison:
- [matplotlib](https://matplotlib.org/devel/coding_guide.html#summary-for-pr-reviewers)
- [scikit-learn](https://scikit-learn.org/stable/developers/contributing.html#code-review-guidelines)
- [Django](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist)

None of the guidelines are set in stone - I just took what I felt was reasonable. Feel free to make changes and propose new ideas. 

Thanks!

melissawm committed:
DOC: Added a PR reviewer guideline document.

bjnath commented:
Thank you for pulling together this much-needed and much-anticipated doc.

***
1. 
And that sets the scene for a suggested opening to the checklist:

* Every PR, good or bad, is an act of generosity. Opening with a positive comment will help the author feel rewarded, and your subsequent remarks may be heard more clearly. You may feel good also.

* Begin if possible with the large issues, so the author knows they've been understood.  Resist the temptation to immediately go line by line, or to open with small pervasive issues.

* You are the face of the project, and NumPy some time ago decided [the kind of project it will be:](https://numpy.org/code-of-conduct/)  open, empathetic, welcoming, friendly, and patient. 

***
2. 
The top paragraph of the guidelines will be read by potential volunteers who are unsure whether they're qualified and don't know where to begin. 

* I would say explicitly "You don't have to be a project member to contribute reviews." 
* Let's give guidance on how people can find PRs that are suited to their skills
* Let's give a profile of who might be a reviewer

***
3. 
Let's put moral guidance (like kindness or "don't let perfect be the enemy of good") in a section separate from items about deliverables ("Is the intended behavior clear under all conditions?").

Also the distinction between "Reviewer checklist" and "Detailed guidelines" isn't clear.  Put yourself in the reader's place: would you know where to find your answer, based on those headings?

I'll take a crack at a possible organization.

bjnath commented:
Also the intended audience isn't clear -- the first lines seem to invite the public, but there's also guidance relevant only to those who (as @charris says) push the button. That might be broken out separately.

bjnath commented:
Should there be a checklist item to ensure tests have been included if there's new or changed functionality?

Qiyu8 commented on doc/source/dev/index.rst:
```diff
@@ -281,6 +286,7 @@ The rest of the story
    Git Basics <gitwash/index>
    development_environment
    development_workflow
+   reviewer-guidelines
```
the file name should be unified with underline.

melissawm committed:
Reorganized content and changed file name.

melissawm commented:
Thanks for the thoughtful comments, @bjnath ! I've updated the PR to reflect that - and changed the document quite a bit. Let me know if this makes more sense. 

Just a comment: I didn't add explicit items about tests because that is part of the main developer guidelines that are linked in the text : 
> When merging contributions, a committer is responsible for ensuring that those meet the requirements outlined in the [Development process guidelines](https://numpy.org/devdocs/dev/index.html#guidelines)

However, if you think they are worth repeating, I'm happy to add them here.

bjnath commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -7,19 +7,48 @@ Reviewer Guidelines
 Reviewing open pull requests (PRs) is very welcome and a valuable way to help
 increase the speed at which the project moves forward. We encourage everyone to
 get involved in the review process; it's also a great way to get familiar with
```
```suggestion
Reviewing open pull requests (PRs) is very welcome and helps
the project move forward. We encourage everyone to
get involved; it's also a great way to get familiar with
```

bjnath commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -7,19 +7,48 @@ Reviewer Guidelines
+Who can be a reviewer?
+======================
+
+You don't have to be a project member to contribute reviews. Consider being a
+reviewer if 
+
+- You have contributed to the NumPy project before, and feel like you can
+  comment on technical or organizational aspects of the PR; 
+- You haven't contributed before, but are an experienced NumPy user, and see a
+  pull request that touches a funcionality that you use often;
```
```suggestion
You don't have to be a project member to contribute reviews. Consider reviewing
if you're an experienced user and the request touches functionality you use often.
```
```

bjnath commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -7,19 +7,48 @@ Reviewer Guidelines
+  comment on technical or organizational aspects of the PR; 
+- You haven't contributed before, but are an experienced NumPy user, and see a
+  pull request that touches a funcionality that you use often;
+- You can contribute comments about content, style, organization or good
+  documentation practices.  
```
Not needed; review topics are discussed in detail below.

bjnath commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -7,19 +7,48 @@ Reviewer Guidelines
+- You can contribute comments about content, style, organization or good
+  documentation practices.  
+
+If we do not know you yet, consider introducing yourself before you start
+reviewing pull requests.
```
Be clearer about what "introducing yourself" means.

bjnath commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -7,19 +7,48 @@ Reviewer Guidelines
+  project it will be <https://numpy.org/code-of-conduct/>`_: open, empathetic,
+  welcoming, friendly and patient. Be `kind
+  <https://youtu.be/tzFWz5fiVKU?t=49m30s>`_ to contributors.
+- Do not let perfect be the enemy of the good, particularly for documentation.
+  If you find yourself making many small suggestions, either open a PR against
+  the original branch, push changes to the contributor branch, or merge the PR
+  and then open a new PR against upstream.
```
* It's not clear how these steps keep perfect from being the enemy of the good or how they move things along.
* Some of these steps are available only to maintainers

bjnath commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -7,19 +7,48 @@ Reviewer Guidelines
+  If you find yourself making many small suggestions, either open a PR against
+  the original branch, push changes to the contributor branch, or merge the PR
+  and then open a new PR against upstream.
+- If you need help writing replies in reviews, check out some `Standard replies
+  for reviewing
+  <https://scikit-learn.org/stable/developers/tips.html#saved-replies>`_.
```
Let's have our own canned set. It's more respectable, it seems easy to do, and it will avoid people copying scikit-learn stuff by accident.

rgommers replied:
+1, but can be done in a follow up PR

bjnath commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -7,19 +7,48 @@ Reviewer Guidelines
-  compatibility breaks were discussed on the `numpy-discussion mailing list
-  <https://mail.python.org/mailman/listinfo/numpy-discussion>`_.
 - Is the intended behavior clear under all conditions (e.g. unexpected inputs
   like empty arrays or nan/inf values)?
```
This is too skimpy. That's ironic for an item that's urging thoroughness. We should give a fuller checklist of cases likely to be overlooked and suggestions on how to think of them generally. The list might be elsewhere and linked to here. 

We should also mention (possibly in a separate bullet) the `hypothesis` tool, for anyone who knows (or may want to learn) how to use it.

rgommers replied:
Empty arrays and inf/nan are the main special cases. Nothing else immediately comes to mind - do you have any suggestions?

mattip replied:
Here are some more thoughts, some may be impractical/not appropriate
- axis or shape arguments should be int and tuples
- scalars and 0-d arrays
- there are techniques to use all the dtypes, which means that new ones (will we ever do another datetime?) will be added to the tests automatically when they are added to NumPy
- tests should hit the error cases in the code

seberg replied:
object arrays are huge special cases (e.g. array containing fractions/decimals).

rgommers replied:
> object arrays are huge special cases (e.g. array containing fractions/decimals).

Yes, but do we really want to require tests with things like object arrays for decimals for all functionality? That seems like an unhealthy idea.

rgommers replied:
+1 to all dtypes and axis int/tuple. the rest seems like overdoing it for a general review recommendation

seberg replied:
@rgommers I thought this was a "checklist" for what reviewers may want to look out for, not a strict requirement?

rgommers replied:
Yes, but inevitably newer reviewers are going to ask for that stuff if it's on this list, and for something obscure like object array of decimals that's almost never desired. So making the list overly long isn't ideal imho.

melissawm replied:
Can one of you explain the technique for testing all dtypes? I'm not familiar with it and don't really know how to explain. I'd say something like:

> Instead of listing specific dtypes when testing, *use technique*. This means that when new dtypes are added to NumPy, the tests will automatically include them.

rgommers replied:
There's no standard way of grabbing a set of dtypes. Tests normally use `pytest.mark.parametrize`. For example:
```
int_types = [np.byte, np.short, np.intc, np.int_, np.longlong]
uint_types = [np.ubyte, np.ushort, np.uintc, np.uint, np.ulonglong]
float_types = [np.half, np.single, np.double, np.longdouble]
cfloat_types = [np.csingle, np.cdouble, np.clongdouble]


class TestArrayFromScalar:

    @pytest.mark.parametrize('t1', int_types + uint_types)
    @pytest.mark.parametrize('t2', int_types + uint_types + [None])
    def test_integers(self, t1, t2):
        return self._do_test(t1, t2)

    @pytest.mark.parametrize('t1', float_types)
    @pytest.mark.parametrize('t2', float_types + [None])
    def test_reals(self, t1, t2):
        return self._do_test(t1, t2)
```

The dtypes is something that requires some thought. Sometimes it uses `float32, float64`, or `"f4", "f8", ">f8",`. A search for `mark.parametrize` will show you a lot of ways of doing this.

rgommers replied:
Actually, this is not the place to explain even `parametrize`. Maybe just add: _"are unusual dtypes tested if a function supports those?"_

bjnath commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -7,19 +7,48 @@ Reviewer Guidelines
-  <https://mail.python.org/mailman/listinfo/numpy-discussion>`_.
 - Is the intended behavior clear under all conditions (e.g. unexpected inputs
   like empty arrays or nan/inf values)?
 - Is the code easy to read and low on redundancy? Should variable names be
```
We don't have to tell reviewers to complain if they can't make sense of the code. But we want them to make constructive suggestions, not just complain, and the list of failures and improvements is much larger than these two items.  Rather than opening the door for vague readability complaints, we should either cover readability comprehensively (possibly with a link), or omit talk of readability and break out the two items (variables and comments) individually

bjnath commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -28,21 +57,33 @@ Reviewer Checklist
 - Does the documentation follow the :ref:`NumPy guidelines<howto-document>`? Are
   the docstrings properly formatted?
 - If it is not obvious from the PR description, add a short explanation of what
-  a branch did to the merge message and, if closing a bug, also add "Closes
```
The advice on merge messages is pertinent only to maintainers.

bjnath commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -28,21 +57,33 @@ Reviewer Checklist
-  a branch did to the merge message and, if closing a bug, also add "Closes
+  a branch did to the merge message and, if closing an issue, also add "Closes
   #123" where 123 is the issue number.
+- For code changes, at least one core team member (those with commit rights)
+  should review and approve all pull requests. If you are the first to review a
```
Not sure of the value of the first sentence. The only way the PR can be committed to master is if someone with commit rights does it.
```suggestion
-  If you are the first to review a
```

rgommers commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -0,0 +1,135 @@
+  the docstrings properly formatted?
+- If it is not obvious from the PR description, add a short explanation of what
+  a branch did to the merge message and, if closing an issue, also add "Closes
+  #123" where 123 is the issue number.
```
minor: `gh-123` would be preferred

rgommers commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -0,0 +1,135 @@
+  <https://help.github.com/articles/reviewing-changes-in-pull-requests/>`_ tool
+  to mark it as such. If you are a subsequent reviewer, please approve the
+  review or, if you have commit rights and think no more review is needed, merge
+  the PR.
```
This seems like a leftover from scikit-image, where they require two reviewers.

melissawm replied:
Actually the way I wanted to write this is that two reviewers are not required, but if someone wants to make suggestions on an already approved PR they can (even with no commit rights). If that's not clear I can rephrase

rgommers replied:
Let me try to propose a rephrase. Like this it seems to encourage a habit I see too often, which is that all important review is done, and then someone comes by and adds one or a couple of nitpicky comments. Which then stalls the merge and is super annoying for the contributor, for no real gain. I want to make clear that's very much not desired.

rgommers replied:
Proposed change to last bullet:
```
- For code changes, at least one maintainer (i.e. someone with commit rights)
  should review and approve a pull request. If you are the first to review a
  PR and approve of the changes use the GitHub `approve review
  <https://help.github.com/articles/reviewing-changes-in-pull-requests/>`_ tool
  to mark it as such. If a PR is straightforward, for example it's a clearly
  correct bug fix, it can be merged straight away. If it's more complex or
  changes public API, please leave it open for at least a couple of days so
  other maintainers get a chance to review.
- If you are a subsequent reviewer on an already approved PR, please use the
  same review method as for a new PR (focus on the larger issues, resist the
  temptation to add only a few nitpicks).  If you have commit rights and think
  no more review is needed, merge the PR.
```

rgommers commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -0,0 +1,135 @@
+  review or, if you have commit rights and think no more review is needed, merge
+  the PR.
+
+Core team members
```
Can we change this to "For maintainers"? I really dislike the term "core team" (implies that some type of activities are more important than others) and want to get rid of it soon.

rgommers commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -0,0 +1,135 @@
+-----------------
+  
+- Make sure all automated CI tests pass before merging a PR, and that the
+  :ref:`documentation builds<building-docs>` without any errors.
```
missing a space between `builds` and `<`

rgommers commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -0,0 +1,135 @@
+  compatibility breaks were discussed on the `numpy-discussion mailing list
+  <https://mail.python.org/mailman/listinfo/numpy-discussion>`_.
+- Backports and trivial additions to finish a PR (really trivial, like a typo or
+  PEP8 fix) can be pushed directly.
```
Let's get rid of this bullet, we don't do that anymore.

rgommers commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -0,0 +1,135 @@
+  done on GitHub, PR page, Files changed tab, Review changes (button on the top
+  right).
+- If you're happy about the current status, mark the pull request as Approved
+  (same way as Changes requested). Alternatively (for core team members): merge
```
another "core team" instance here

rgommers commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -0,0 +1,135 @@
+own machine so that you can play with it locally. One way to do this is to
+navigate to the NumPy root directory in the terminal and enter::
+  
+  git fetch upstream pull/PULL_REQUEST_ID/head:NEW_BRANCH_NAME
```
Let's replace this bit with `gh` or `hub` CLI usage (or get rid of it).

rgommers replied:
Oh wait, that's below. I think that's so far superior that we should just delete this old hacky method.

rgommers replied:
this apparently does work, learned something. resolving.

rgommers reviewed (commented):
Thanks @melissawm, this looks really good! I think we should merge this soon. I made a few smaller comments inline. The main thing I'm still missing is two things _not_ to do: being too nitpicky on style/taste things, and on grammar. Instead, we should merge things quicker when all important concerns are addressed, and if the reviewer really wants to fix the last little nit then either push a commit directly or make a follow-up PR themselves.

melissawm commented:
Thanks for the thoughtful comments, all! I'll work on a revised version as soon as possible.

rgommers commented:
> The top paragraph of the guidelines will be read by potential volunteers who are unsure whether they're qualified and don't know where to begin.

This is a fair comment. Let's not make things too open. If someone is new to reviewing code or for some other reason is unsure, most likely they're not the right person to start reviewing. Most non-maintainers who do end up making constructive reviewer contributions are domain experts (e.g. linalg, fft) or experienced maintainers of some other project.

bjnath commented:
> Let's not make things too open.

So maybe a sentence clarifying roles, so everyone knows where and whether they fit in:

> Reviews can come from outside the NumPy team --  we welcome contributions from domain experts (for instance,  linalg, fft) or maintainers who work on another project. You do not need to be a maintainer to review.
> 
> A "maintainer" is a NumPy team member with permission to merge a PR into the NumPy master branch ("commit rights"). Maintainers are listed [here](https://numpy.org/gallery/team.html).
>

rgommers commented:
That looks good @bjnath, let's just leave out the link to the gallery since that's usually going to be (a) out of sync, and (b) not include everyone with commit rights such as previous maintainers who have been inactive for a long time.

melissawm committed:
Incorporated several suggestions from PR review.

melissawm commented:
I think I addressed everything. Let me know if there's anything I missed!

eric-wieser commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -0,0 +1,118 @@
+   - Are unusual `dtypes` tested if a function supports those?
+- Should variable names be improved for clarity or consistency?
+- Should comments be added, or rather removed as unhelpful or extraneous?
+- Does the documentation follow the :ref:`NumPy guidelines<howto-document>`? Are
+  the docstrings properly formatted?
```
On the same note, can we add a reference to the code style guidelines here? https://numpy.org/doc/stable/dev/index.html#stylistic-guidelines (but with an internal sphinx link)

rgommers replied:
done, resolving.

melissawm committed:
Added link to stylistic guidelines for code.

rgommers commented on doc/source/dev/reviewer_guidelines.rst:
```diff
@@ -0,0 +1,119 @@
+   - Are unusual `dtypes` tested if a function supports those?
+- Should variable names be improved for clarity or consistency?
+- Should comments be added, or rather removed as unhelpful or extraneous?
+- Does the documentation follow the :ref:`NumPy guidelines<howto-document>`? Are
```
still an error in the reference here, needs a space

rgommers replied:
or maybe this does work, since you're doing it in multiple place. let me go look at the html

rgommers commented:
Looks great, in it goes. Thanks Melissa!
